// parcel/lib/transforms/posthtml.js
'use strict';

let parse = (() => {
  var _ref = _asyncToGenerator(function* (code, asset) {
    var config = yield getConfig(asset);
    if (!config) {
      config = {};
    }
    config = Object.assign({ lowerCaseAttributeNames: true }, config);
    return posthtmlParse(code, config);
  });

  return function parse(_x, _x2) {
    return _ref.apply(this, arguments);
  };
})();

let transform = (() => {
  var _ref2 = _asyncToGenerator(function* (asset) {
    let config = yield getConfig(asset);
    if (!config) {
      return;
    }

    yield asset.parseIfNeeded();
    let res = yield posthtml(config.plugins).process(asset.ast, config);

    asset.ast = res.tree;
    asset.isAstDirty = true;
  });

  return function transform(_x3) {
    return _ref2.apply(this, arguments);
  };
})();

let getConfig = (() => {
  var _ref3 = _asyncToGenerator(function* (asset) {
    let config = yield asset.getConfig(['.posthtmlrc', '.posthtmlrc.js', 'posthtml.config.js'], {
      packageKey: 'posthtml'
    });
    if (!config && !asset.options.minify) {
      return;
    }

    config = Object.assign({}, config);
    const plugins = config.plugins;
    if (typeof plugins === 'object') {
      const depConfig = {
        addDependencyTo: {
          addDependency: function addDependency(name) {
            return asset.addDependency(name, { includedInParent: true });
          }
        }
      };
      Object.keys(plugins).forEach(function (p) {
        return Object.assign(plugins[p], depConfig);
      });
    }
    config.plugins = yield loadPlugins(plugins, asset.name);
    config.skipParse = true;
    return config;
  });

  return function getConfig(_x4) {
    return _ref3.apply(this, arguments);
  };
})();

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const loadPlugins = require('../utils/loadPlugins');
const posthtml = require('posthtml');
const posthtmlParse = require('posthtml-parser');

module.exports = { parse, transform };
